desc/objectives/_omnigenity.py

```diff
@@ -540,6 +541,352 @@ def compute(self, params, constants=None):
+        associated data are re-computed at every iteration (Default).
+    field_fixed: bool, optional
+        Whether the OmnigenousField `field` is fixed or not.
+        If True, the field is fixed and its values are precomputed, which saves on
+        computation time during optimization and self.things = [eq] only.
+        If False, the field is allowed to change during the optimization and its
+        associated data are re-computed at every iteration (Default).
+    name : str
+        Name of the objective function.
+
+    """
+
+    _coordinates = "rtz"
+    _units = "(T)"
+    _print_value_fmt = "Omnigenity error: {:10.3e} "
+
+    def __init__(
+        self,
+        eq,
+        field,
+        target=None,
+        bounds=None,
+        weight=1,
+        normalize=True,
+        normalize_target=True,
+        loss_function=None,
+        deriv_mode="fwd",  # FIXME: get it working with rev mode, set default to auto
+        eq_grid=None,
+        field_grid=None,
+        M_booz=None,
+        N_booz=None,
+        eta_weight=1,
+        eq_fixed=False,
+        field_fixed=False,
+        name="omnigenity",
+    ):
+        if target is None and bounds is None:
+            target = 0
+        self._eq = eq
+        self._field = field
+        self._eq_grid = eq_grid
+        self._field_grid = field_grid
+        self.helicity = field.helicity
+        self.M_booz = M_booz
+        self.N_booz = N_booz
+        self.eta_weight = eta_weight
+        self._eq_fixed = eq_fixed
+        self._field_fixed = field_fixed
+        if not eq_fixed or field_fixed:
+            things = [eq, field]
+        elif eq_fixed and not field_fixed:
+            things = [field]
+        elif field_fixed and not eq_fixed:
+            things = [eq]
+        else:
+            raise ValueError("Cannot fix both the eq and field.")
+        super().__init__(
+            things=things,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        if self._eq_fixed:
+            eq = self._eq
+            field = self.things[0]
+        elif self._field_fixed:
+            eq = self.things[0]
+            field = self._field
+        else:
+            eq = self.things[0]
+            field = self.things[1]
+
+        M_booz = self.M_booz or 2 * eq.M
+        N_booz = self.N_booz or 2 * eq.N
+
+        if self._eq_grid is None:
+            eq_grid = LinearGrid(M=2 * M_booz, N=2 * N_booz, NFP=eq.NFP, sym=False)
```

Comment:
if the user supplies a field grid but not an eq grid, we should have the eq grid default to the same rho value as the user supplied field grid, and vice versa.